src/compiler/checker.ts

```diff
@@ -26430,14 +26420,14 @@ namespace ts {
             }
         }
 
-        function checkFunctionExpressionOrObjectLiteralMethod(node: FunctionExpression | MethodDeclaration, checkMode?: CheckMode): Type {
+        function checkFunctionExpressionOrObjectLiteralMethod(node: FunctionExpression | ArrowFunction | MethodDeclaration, checkMode?: CheckMode): Type {
             Debug.assert(node.kind !== SyntaxKind.MethodDeclaration || isObjectLiteralMethod(node));
             checkNodeDeferred(node);
 
             // The identityMapper object is used to indicate that function expressions are wildcards
             if (checkMode && checkMode & CheckMode.SkipContextSensitive && isContextSensitive(node)) {
                 // Skip parameters, return signature with return type that retains noncontextual parts so inferences can still be drawn in an early stage
-                if (!getEffectiveReturnTypeNode(node) && hasContextSensitiveReturnExpression(node)) {
+                if (!getEffectiveReturnTypeNode(node) && node.kind === SyntaxKind.ArrowFunction && node.parameters.length === 0) {

```

Comment:
This seems wrong; it does not admit annotated parameters; I see no reason why a list of annotated parameters would be treated differently than an empty parameter list when it comes to contextual-ness. Moreover, this inverts the check; rather than seeing if the return is context sensitive, this is checking if the parameters _aren't_ (almost). We have no reason to do the return type special-casing below if the return _isn't_ context sensitive (it's extra work we shouldn't do since it'll be repeated in later inference phases), so we should still be checking that the return expression is context sensitive, even if we pile on additional constraints.

Reply:
Yeah, we should permit functions with no contextually sensitive parameters.